festim/hydrogen_transport_problem.py

```diff
@@ -503,11 +535,14 @@ def create_formulation(self):
                 self.formulation += ((u - u_n) / self.dt) * v * self.dx(vol.id)
 
                 # add sources
-                # TODO implement this
-                # for source in self.sources:
-                #     # f = Constant(my_mesh.mesh, (PETSc.ScalarType(0)))
-                #     if source.species == spe:
-                #         formulation += source * v * self.dx
+                for source in self.sources:
+                    for source_spe in source.species:
+                        for source_vol in source.volume:
+                            if source_spe == spe and source_vol == vol:
+                                self.formulation -= (
+                                    source.value_fenics * v * self.dx(vol.id)
+                                )
```

Comment:
This can go outsie the main species loop and just iterate through the the sources. Something like:

```python
for source in self.sources:
    for source_spe in source.species:
        v = source_spe.test_function
        for source_vol in source.volume:
            self.formulation -= (
                source.value_fenics * v * self.dx(source_vol.id)
            )
```